### PR TITLE
326 add tensorrt test scripts [skip gpu]

### LINKED ISSUE #326: Enhance CI tests for TensorRT supported bundles
Currently, all bundles use the same pipeline to do pre-merge tests. For each changed bundle (within a pull request), its required libs (such as MONAI version, Torch version, ...) will be extracted from `metadata.json`, and installed with `pipenv` during the testing workflow.

However, to enable TensorRT for a part of bundles (see also: https://github.com/Project-MONAI/model-zoo/issues/324), we need to add related tests. Install TensorRT requires more works than `pipenv install`, thus we decided to use a MONAI docker container (https://docs.monai.io/en/latest/installation.html#from-dockerhub) to do the test.

A limitation is that we need to align the libs (versions) used for all TensorRT supported bundles. As the first stage, three bundles will support TensorRT (see also: https://github.com/Project-MONAI/model-zoo/pull/325), and they will unify MONAI and Torch versions.

The expected new CI tests logic is like:

Screen Shot 2023-03-15 at 10 56 53 AM

As shown in the figure, we will support two kinds of tests:

1. For each pull request, if a TensorRT supported bundle is changed, it will trigger a corresponding TensorRT test within MONAI docker container (latest LTS version).
2. Periodically (such as weekly) TensorRT tests will be executed on all supported bundles.

### PR DESCRIPTION
**Pull request by yiheng-wang-nv**
Fixes #326 .

### Description
This PR adds tensorrt related checks. They will be called in blossom ci.
In addition, this PR also simplifies ci tests in order to efficiently develop MONAI 1.2 bundles. Added workarounds are expected to be removed when merging `dev-1.2` branch into `dev`.

After this PR is merged, @YanxuanLiu will help to build a pipeline which enables us to run periodically tests on all TensorRT bundles.

time to test for each bundle (one A100 80G):

- `spleen_ct_segmentation`
    - fp 16: 12 min
    - fp 32: 4 min
- `pathology_tumor_detection`
    - fp 16: 30s
    - fp 32: 16s
- `endoscopic_tool_segmentation`
    - fp 16: 5 min
    - fp 32: 3 min

### Status
**Ready**

### Please ensure all the checkboxes:

- [x] Codeformat tests passed locally by running `./runtests.sh --codeformat`.
- [ ] In-line docstrings updated.
- [ ] Please ensure the naming rules in config files meet our requirements (please refer to: `CONTRIBUTING.md`).
- [ ] Ensure versions of packages such as `monai`, `pytorch` and `numpy` are correct in `metadata.json`.
- [ ] Descriptions should be consistent with the content, such as `eval_metrics` of the provided weights and TorchScript modules.
- [ ] Files larger than 25MB are excluded and replaced by providing download links in `large_file.yml`.
- [ ] Avoid using path that contains personal information within config files (such as use `/home/your_name/` for `"bundle_root"`).

### CODE REVIEW

File: ci/verify_tensorrt.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import argparse
+
+from utils import get_changed_bundle_list
+
+
+def verify_tensorrt(changed_dirs):
+    """
+    This function is used to verify if the checkpoint is able to export into TensorRT module.
+    This function will be updated after the following PR is merged:
+    https://github.com/Project-MONAI/MONAI/pull/5986/files
```
Comment (yiheng-wang-nv) [inline ci/verify_tensorrt.py]: mark here. Will update the actual test function after the API is ready.

File: ci/verify_tensorrt.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        ckpt_file=os.path.join(bundle_path, "models/model.pt"),
+        meta_file=os.path.join(bundle_path, "configs/metadata.json"),
+        config_file=os.path.join(bundle_path, config_file),
+        precision=precision,
+        bundle_root=bundle_path,
+        dynamic_batchsize=(1, 1, 1),
+    )
+
+    _ = torch.jit.load(trt_model_path)
+
```
Comment (yiheng-wang-nv) [inline ci/verify_tensorrt.py:45]: Hi @binliunls , could you please help to check if the trt test here is enough?
As for the `dynamic_batchsize`, I hard coded to use (1, 1, 1). 
The function will check:
1. if bundle weights can be exported into trt model.
2. if the exported trt model can be loaded.

Comment (binliunls) [reply ci/verify_tensorrt.py:45]: Hi @yiheng-wang-nv ,
      The `dynamic_batchsize` support the None input with a warning.  Maybe you can just remove the `dynamic_batchsize` line.

Hi @Nic-Ma ,
       Do you think if it is necessary that we add a speadup ratio > 1 check for model computation?

Thanks,
Bin

Comment (Nic-Ma) [reply ci/verify_tensorrt.py:45]: Hi @binliunls ,

Let's focus on the functionality tests for the first step, we can add more speed tests later.

Thanks.

**Comment by Nic-Ma** [comment]
Let's drop the `dev-1.2` branch and use `dev` branch with `monai 1.2.0rc2` directly.

Thanks.

File: ci/verify_tensorrt.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    for bundle in include_verify_tensorrt_list:
+        print(f"start verifying bundle {bundle}.")
+        bundle_path = os.path.join(models_path, bundle)
+        net_id, inference_file_name = "network_def", _find_bundle_file(
+            os.path.join(bundle_path, "configs"), "inference"
+        )
+        config_file = os.path.join("configs", inference_file_name)
+        for precision in ["fp32", "fp16"]:
+            verify_tensorrt(bundle_path=bundle_path, net_id=net_id, config_file=config_file, precision=precision)
+            print(f"export bundle {bundle} weights into TensorRT module with precision {precision} successfully.")
```
Comment (Nic-Ma) [inline ci/verify_tensorrt.py]: What's the behavior if failed this test?
Just raise an exception?

Thanks.

File: ci/bundle_custom_data.py [hunk 4]
```diff
@@ -30,3 +30,7 @@
     "pathology_tumor_detection",
     "mednist_reg",
 ]
+
+# This list is used for our CI tests to determine whether a bundle supports TensorRT. Related
+# test will be employed for bundles in the list.
+include_verify_tensorrt_list = []
```
Comment (Nic-Ma) [inline ci/bundle_custom_data.py:36]: Why the list is empty?

Thanks.

Comment (yiheng-wang-nv) [reply ci/bundle_custom_data.py:36]: for current dev branch, no bundle supports tensorrt. Require users to add supported bundles into the list within their PRs.

File: ci/verify_tensorrt.py (hunk 3 above)
Comment (yiheng-wang-nv) [reply ci/verify_tensorrt.py]: Thanks Nic for the comments. I added error messages and could you please help to recheck it?

File: ci/bundle_custom_data.py (hunk 4 above)
Comment (Nic-Ma) [reply ci/bundle_custom_data.py:36]: OK, I think @binliunls 's PR can add the 3 bundles to this list.

Thanks.

File: ci/verify_tensorrt.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            meta_file=os.path.join(bundle_path, "configs/metadata.json"),
+            config_file=os.path.join(bundle_path, config_file),
+            precision=precision,
+            bundle_root=bundle_path,
+        )
+    except Exception as e:
+        print(f"'trt_export' failed with error: {e}")
+        raise
+    try:
+        _ = torch.jit.load(trt_model_path)
```
Comment (Nic-Ma) [inline ci/verify_tensorrt.py]: I think this line can be simplified to:
```py
torch.jit.load(trt_model_path)
```
No need the "_ =".

Thanks.

**Review by Nic-Ma — APPROVED** [review]

File: ci/verify_tensorrt.py (hunk 3 above)
Comment (Nic-Ma) [reply ci/verify_tensorrt.py]: OK, thanks for the quick update.

File: ci/verify_tensorrt.py (hunk 5 above)
Comment (yiheng-wang-nv) [reply ci/verify_tensorrt.py]: thanks Nic, I just addressed it.

**Pull request merged**